File: src/main/java/org/raku/comma/psi/RakuModuleReference.java
package org.raku.comma.psi;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReferenceBase;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.util.IncorrectOperationException;
import org.raku.comma.psi.stub.index.RakuStubIndexKeys;
import org.raku.comma.psi.stub.index.ProjectModulesStubIndex;
import org.raku.comma.utils.RakuModuleListFetcher;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class RakuModuleReference extends PsiReferenceBase<RakuModuleName> {
    public RakuModuleReference(@NotNull RakuModuleName moduleName) {
        super(moduleName, new TextRange(0, moduleName.getTextLength()));
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        Project project = this.getElement().getProject();
        Collection<RakuFile> elements = StubIndex.getElements(RakuStubIndexKeys.PROJECT_MODULES, getValue(), project, GlobalSearchScope.allScope(project), RakuFile.class);
        return  elements.isEmpty() ? null : elements.iterator().next();
    }

    @Override
    public Object @NotNull [] getVariants() {
        Project project = this.getElement().getProject();
        Collection<String> projectModules = ProjectModulesStubIndex.getInstance().getAllKeys(project);
        List<String> reallyInThisProject = new ArrayList<>();
        projectModules.forEach(module -> {
            var index = ProjectModulesStubIndex.getInstance();
            Collection<RakuFile> matching = StubIndex.getElements(index.getKey(),
                                                                  module,
                                                                  project,
                                                                  GlobalSearchScope.projectScope(project),
                                                                  RakuFile.class);
            if (! matching.isEmpty()) {
                reallyInThisProject.add(module);
            }
        });

        if (RakuModuleListFetcher.isReady()) {
            reallyInThisProject.addAll(RakuModuleListFetcher.getProvides(myElement.getProject()));
        }
        reallyInThisProject.addAll(RakuModuleListFetcher.PREINSTALLED_MODULES);
        reallyInThisProject.addAll(RakuModuleListFetcher.PRAGMAS);
        return reallyInThisProject.toArray();
    }

    @Override
    public PsiElement handleElementRename(@NotNull String newElementName) throws IncorrectOperationException {
        RakuModuleName name = getElement();
        return name.setName(newElementName);
    }

    @Override
    public PsiElement bindToElement(@NotNull PsiElement element) throws IncorrectOperationException {
        // Our RakuFile, so can calculate new path
        if (element instanceof RakuPsiElement) {
            getElement().setName(((RakuPsiElement)element).getEnclosingPerl6ModuleName());
        }
        return element;
    }
}

File: src/main/java/org/raku/comma/psi/RakuModuleReference.kt
package org.raku.comma.psi

import com.intellij.openapi.components.service
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager
import com.intellij.psi.PsiReferenceBase
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import com.intellij.psi.util.CachedValue
import com.intellij.psi.util.CachedValueProvider
import com.intellij.psi.util.CachedValuesManager
import com.intellij.psi.util.PsiModificationTracker
import com.intellij.util.IncorrectOperationException
import org.raku.comma.psi.stub.index.ProjectModulesStubIndex
import org.raku.comma.psi.stub.index.RakuStubIndexKeys
import org.raku.comma.services.project.RakuDependencyService
import org.raku.comma.services.RakuServiceConstants
import java.util.function.Consumer

class RakuModuleReference(moduleName: RakuModuleName) :
    PsiReferenceBase<RakuModuleName?>(moduleName, TextRange(0, moduleName.textLength)) {
    private val project = moduleName.project
    private val psiManager = PsiManager.getInstance(project)
    private var cachedFile: CachedValue<PsiFile>? = null

    override fun resolve(): PsiElement? {
        if (cachedFile?.value == null) {
            val tryAgain = doResolve() as? PsiFile ?: return null
            cachedFile = createCachedValue(tryAgain)
        }
        return cachedFile?.value
    }

    private fun doResolve(): PsiElement? {
        val keys = StubIndex.getElements(
            RakuStubIndexKeys.PROJECT_MODULES,
            value,
            project,
            GlobalSearchScope.projectScope(project),
            RakuFile::class.java
        )
        val rakuFile = if (keys.isNotEmpty()) keys.first() else return resolveExternal()

        return psiManager.findFile(rakuFile.virtualFile)
    }

    private fun resolveExternal(): PsiFile? {
        return project.service<RakuDependencyService>().provideToPsiFile(value)
    }

    private fun createCachedValue(file: PsiFile): CachedValue<PsiFile> {
        return CachedValuesManager.getManager(project).createCachedValue {
            CachedValueProvider.Result(file, PsiModificationTracker.MODIFICATION_COUNT)
        }
    }

    override fun getVariants(): Array<Any> {
        val project = this.element.project
        val projectModules = ProjectModulesStubIndex.getInstance().getAllKeys(project)
        val reallyInThisProject: MutableList<String> = ArrayList()
        projectModules.forEach(Consumer { module: String ->
            val index = ProjectModulesStubIndex.getInstance()
            val matching = StubIndex.getElements(
                index.key,
                module,
                project,
                GlobalSearchScope.projectScope(project),
                RakuFile::class.java
            )
            if (!matching.isEmpty()) {
                reallyInThisProject.add(module)
            }
        })

        reallyInThisProject.addAll(project.service<RakuDependencyService>().allProvides())
        reallyInThisProject.addAll(RakuServiceConstants.PREINSTALLED_MODULES)
        reallyInThisProject.addAll(RakuServiceConstants.PRAGMAS)

        return reallyInThisProject.toTypedArray()
    }

    @Throws(IncorrectOperationException::class)
    override fun handleElementRename(newElementName: String): PsiElement {
        val name = element
        return name.setName(newElementName)
    }

    @Throws(IncorrectOperationException::class)
    override fun bindToElement(element: PsiElement): PsiElement {
        // Our RakuFile, so can calculate new path
        if (element is RakuPsiElement) {
            getElement().setName(element.enclosingRakuModuleName)
        }
        return element
    }
}
